lib/puppet/provider/service/src.rb

```diff
@@ -63,6 +65,23 @@ def disable
     rmitab(@resource[:name])
   end
 
+  def stop
+    ucommand(:stop)
+
+    # Wait for the service to finish stopping
+    begin
+      Timeout.timeout(60) do
+        loop do
+          status = self.status
+          break if status == :stopped
+          sleep(1)
+        end
+      end
+    rescue Timeout::Error
+      raise Puppet::Error.new("Timed out waiting for #{@resource[:name]} to stop")
```

Comment:
The transaction should mark the resource as failed if the provider throws exceptions. Are other providers explicitly failing their resources?